calibr8/tests.py

```diff
@@ -941,6 +1054,73 @@ def test_likelihood(self):
         assert x_dense[numpy.argmax(actual)] == 2
         pass
 
+    @pytest.mark.xfail(reason="Draft test case, see https://github.com/JuBiotech/calibr8/issues/15")
+    def test_likelihood_nobroadcasting_fallback(self):
+        class _TestSwitchableBroadcastingModel(calibr8.CalibrationModel, calibr8.NormalNoise):
+            def __init__(self):
+                self.x_shapes = []
+                self.can_broadcast = False
+                super().__init__(independent_key="I", dependent_key="D", theta_names="i,s,sd".split(","), ndim=1)
+
+            def predict_dependent(self, x, *, theta=None):
+                if theta is None:
+                    theta = self.theta_fitted
+                i, s, sd = self.theta_fitted
+                # This part broadcasts just fine
+                x = numpy.array(x)
+                mu = i + x * s
+                return mu, sd
+
+            def loglikelihood(self, *, y, x, **kwargs):
+                # This overrides the native CalibrationModel.loglikelihood with one
+                # that can be externally set to return non-broadcasted results.
+                LL_broadcasted = super().loglikelihood(y=y, x=x, **kwargs)
+                if not self.can_broadcast:
+                    return LL_broadcasted.sum()
+                return LL_broadcasted
+
+        cm = _TestSwitchableBroadcastingModel()
+        cm.theta_fitted = [0.5, 0.6, 0.7]
+
+        X = numpy.array([1, 2, 3])
+        Y = [1.1, 1.7, 2.3]
+        y_obs = [1.0, 1.7, 2.4]
+
+        # Check the prediction of the test model
+        for x, y in zip(X, Y):
+            assert cm.predict_dependent(x) == (y, 0.7)
+
+        # The predict_dependent can broadcast
+        mu, sd = cm.predict_dependent(X)
+        numpy.testing.assert_array_equal(mu, Y)
+        assert sd == 0.7
+
+        # Test the switching between can_broadcast modes
+        cm.can_broadcast = True
+        LL = cm.loglikelihood(x=X[..., None], y=y_obs)
+        assert numpy.shape(LL) == (3,)
+
+        cm.can_broadcast = False
+        LL = cm.loglikelihood(x=X[..., None], y=y_obs)
+        assert numpy.shape(LL) == ()
+
+        # The CalibrationModel.likelihood should give the same results either way.
+        cm.can_broadcast = True
+        L_broadcasted = cm.likelihood(x=X[..., None], y=y_obs, scan_x=True)
+        cm.can_broadcast = False
+        L_looped = cm.likelihood(x=X[..., None], y=y_obs, scan_x=True)
+
+        numpy.testing.assert_array_equal(L_broadcasted, L_looped)
```

Comment:
Here you check that in case of scan x, it doesn't matter if you broadcast or not because the function evaluates at each point anyway? Sorry, I think my brain needs to wrap around this once more.

Reply:
Yeah, this is checking the result from the fallback that does the list comprehension.

Note that this test is marked as XFAIL and will be refactored when working on #15

Reply:
Okay, then I leave it for now :)